src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1213 @@
+                              bhh_bytes[24],    bhh_bytes[25],      bhh_bytes[26],      bhh_bytes[27],
+                              bhh_bytes[28],    bhh_bytes[29],      bhh_bytes[30],      bhh_bytes[31]);
+
+        TrieFileStorage::block_dir(dir_path, bhh).join(bhh_name)
+    }
+
+    pub fn block_path_tmp(dir_path: &String, bhh: &BlockHeaderHash) -> PathBuf {
+        // it looks awkward, but it's waaaay faster than just doing to_hex()
+        let bhh_bytes = bhh.as_bytes();
+        let bhh_name = format!("{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}{:02x}.tmp",
+                              bhh_bytes[0],     bhh_bytes[1],       bhh_bytes[2],       bhh_bytes[3],
+                              bhh_bytes[4],     bhh_bytes[5],       bhh_bytes[6],       bhh_bytes[7],
+                              bhh_bytes[8],     bhh_bytes[9],       bhh_bytes[10],      bhh_bytes[11],
+                              bhh_bytes[12],    bhh_bytes[13],      bhh_bytes[14],      bhh_bytes[15],
+                              bhh_bytes[16],    bhh_bytes[17],      bhh_bytes[18],      bhh_bytes[19],
+                              bhh_bytes[20],    bhh_bytes[21],      bhh_bytes[22],      bhh_bytes[23],
+                              bhh_bytes[24],    bhh_bytes[25],      bhh_bytes[26],      bhh_bytes[27],
+                              bhh_bytes[28],    bhh_bytes[29],      bhh_bytes[30],      bhh_bytes[31]);
+
+        TrieFileStorage::block_dir(dir_path, bhh).join(bhh_name)
+    }
+
+    fn cached_block_path(&mut self, bhh: &BlockHeaderHash) -> PathBuf {
+        let (p, miss) = match self.block_path_cache.get(bhh) {
+            Some(ref p) => {
+                ((*p).clone(), false)
+            }
+            None => {
+                (TrieFileStorage::block_path(&self.dir_path, bhh), true)
+            }
+        };
+        
+        if miss {
+            self.block_path_cache.insert(bhh.clone(), p.clone());
+        }
+        p
+    }
+
+    pub fn read_block_parent(dir_path: &String, bhh: &BlockHeaderHash) -> Result<BlockHeaderHash, Error> {
+        let block_path = TrieFileStorage::block_path(dir_path, bhh);
+        let mut fd = fs::OpenOptions::new()
+                    .read(true)
+                    .write(false)
+                    .open(&block_path)
+                    .map_err(|e| {
+                        if e.kind() == io::ErrorKind::NotFound {
+                            error!("File not found: {:?}", &block_path);
+                            Error::NotFoundError
+                        }
+                        else {
+                            Error::IOError(e)
+                        }
+                    })?;
+
+        let mut hashbuf = Vec::with_capacity(TRIEHASH_ENCODED_SIZE);
+        fseek(&mut fd, 0)?;
+        read_hash_bytes(&mut fd, &mut hashbuf)?;
```

Comment:
This function looks like it creates three buffers -- `hashbuf`, `hashbuf_slice`, and within `read_hash_bytes`, it creates another `hashbytes`.

This happens a lot when `read_hash_bytes` is called: I suspect we could eliminate the need for `fast_extend_from_slice` on a majority of reads by changing the signature (or creating a new function) for `read_hash_bytes` to:

```rust
pub fn read_hash_bytes<F: Read + Seek>(f: &mut F, buf: &mut [u8; 32]) -> Result<(), Error> {
    f.read(buf)
        .map_err(Error::IOError)?;
    
    Ok(())
}
```

Reply:
The reason it's implemented this way is because `read_hash_bytes()` is called many times in succession to read all of the hashes for the children of a node (i.e. via the `read_node_hash_bytes()` method, which in turn is called via `Trie::get_children_hashes_bytes()`).  The call is meant to _append_ the bytes to the given buffer, which `fast_extend_from_slice()` accelerates by way of avoiding a `malloc()`.

`read_block_parent()` is only called when loading the fork table into memory, which happens once one the peer initializes.  It's not great that it's inefficient, but it's not a huge performance penalty either.

Reply:
Ah, I see now. The way they deal with this in the Rust libraries, is just by always using slice pointers. That way, the caller provides the piece of the slice that should be filled using the slice notation `&T[from...to]`. This would require a fair amount of refactoring that I think we should do eventually, since it would obviate the need for `fast_extend_from_slice` (removing an `unsafe` block, which is always a win, no matter how safe the code inside may actually be), and likely give us some performance benefits to, but I don't think it should hold back this PR.